### ISSUE TITLE
--noImplicitThis error is inconsistent

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#25526]
**TypeScript Version:**  3.0.0-dev.20180707

**Code**

```ts
declare function f(a: any): void;
f(function() {
	this.m();
});

f({
	callback: function() {
		this.m();
	}
});
```

**Expected behavior:**

With `--noImplicitThis`, errors in both cases or neither case.

**Actual behavior:**

Only error in the first case.

**Comment by 0xorial** [comment]
Is it the same issue as [here](https://stackoverflow.com/questions/52782824/why-typescript-compiler-ignores-noimplicitthis-option-in-my-example?noredirect=1#comment92487174_52782824)?

**Comment by andy-ms** [comment]
That situation seems even worse because you never declared an `any` anywhere. But it would be likely to be fixed when this one is, so I'll make it a part of this issue.
```ts
class x {
    public variable: number = 0;
    public x() {
        const c = [];
        c.push({
            xxx () {
                console.log(this.variable); // Should error
            }
        });
    }
}
```

**Comment by 0xorial** [comment]
Well, my problem is that I was migrating old code which has a whole bunch of constructs of this kind and I was expecting this option to catch those cases, so I wasn't very careful to review them myself. 

I looked through history and it seems that there was a relevant test, but it was removed by [this commit](https://github.com/Microsoft/TypeScript/pull/14141/commits/e512376b0c5ae25d0e66b4c92bc6618cbd262c95).

BTW, did I understand correctly that files reference/noImplicitThisObjectLiterals.errors.txt and reference/noImplicitThisObjectLiterals.js should have been deleted as well?

**Comment by andy-ms** [comment]
> BTW, did I understand correctly that files reference/noImplicitThisObjectLiterals.errors.txt and reference/noImplicitThisObjectLiterals.js should have been deleted as well?

Right. Any test output file with no associated test file should be deleted. PRs accepted.

**Comment by 0xorial** [comment]
> That situation seems even worse because you never declared an any anywhere. 
After some more investigation, I realized that it is not exactly correct. "Array.push" has first parameter of type any[] (ish).

**Comment by andy-ms** [comment]
`Array.push` should have a parameter of type `T` where `T` is the array's element type. With `--noImplicitAny` enabled, if the array didn't have a declared type we should be inferring one -- not just making it `any`.
